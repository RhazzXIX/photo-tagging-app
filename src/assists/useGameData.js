import { useState, useEffect } from "react";
import { database, storage } from "./fbConfig";
import { getDownloadURL, ref } from "firebase/storage";
import { getDoc, doc } from "firebase/firestore";

export default function useGameData(props) {
  const [firstMap, setFirstmap] = useState({});
  const [scndMap, setScndMap] = useState({});
  const [charSelection1, setCharSelection1] = useState([]);
  const [charSelection2, setCharSelection2] = useState([]);

  const selection1Ref = doc(database, "game", "anime");
  const selection2Ref = doc(database, "game", "game");

  const getData = async () => {
    const firstMapData = {};
    const scndMapData = {};
    const char1 = {};
    const char2 = {};
    const char3 = {};
    const char4 = {};
    const char5 = {};
    const char6 = {};

    const data1 = await getDoc(selection1Ref).then((response) =>
      response.data()
    );
    firstMapData.name = data1.map.name;
    await getDownloadURL(ref(storage, data1.map.imgRef)).then(
      (url) => (firstMapData.url = url)
    );
    setFirstmap(firstMapData);
    char1.name = data1.char1.name;
    await getDownloadURL(ref(storage, data1.char1.imgRef)).then(
      (url) => (char1.url = url)
    );
    char2.name = data1.char2.name;
    await getDownloadURL(ref(storage, data1.char2.imgRef)).then(
      (url) => (char2.url = url)
    );
    char3.name = data1.char3.name;
    await getDownloadURL(ref(storage, data1.char3.imgRef)).then(
      (url) => (char3.url = url)
    );
    setCharSelection1([char1, char2, char3]);

    const data2 = await getDoc(selection2Ref).then((response) =>
      response.data()
    );
    scndMapData.name = data2.map.name;
    await getDownloadURL(ref(storage, data2.map.imgRef)).then(
      (url) => (scndMapData.url = url)
    );
    setScndMap(scndMapData);
    char4.name = data2.char1.name;
    await getDownloadURL(ref(storage, data2.char1.imgRef)).then(
      (url) => (char4.url = url)
    );
    char5.name = data2.char2.name;
    await getDownloadURL(ref(storage, data2.char2.imgRef)).then(
      (url) => (char5.url = url)
    );
    char6.name = data2.char3.name;
    await getDownloadURL(ref(storage, data2.char3.imgRef)).then(
      (url) => (char6.url = url)
    );
    setCharSelection2([char4, char5, char6]);
  };

  useEffect(() => {
    getData();
  }, []);

  return { firstMap, scndMap, charSelection1, charSelection2 };
}
