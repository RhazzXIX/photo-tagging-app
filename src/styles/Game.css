main#game {
  flex: auto;
  background-color: #2b2c5a;
  min-width: 1840px;
  min-height: 1142px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-evenly;
  position: relative;
}

section#loading {
  position: absolute;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: white;
  z-index: 2;
}

main#game > header {
  position: sticky;
  top: 0;
  background-color: rgba(255, 255, 255, 0.5);
  padding: 4px;
  width: 100%;
  display: grid;
  grid-template-columns: 1fr;
  grid-row: 1fr;
  z-index: 1;
}

main#game > header div.timer {
  grid-area: 1 / 1 / 2 / 2;
  display: flex;
  align-items: center;
  gap: 4px;
  color: #2b2c5a;
  font-size: 1.8rem;
  font-weight: 700;
  margin-left: 20px;
}

main#game > header ul {
  grid-area: 1 / 1 / 2 / 2;
  display: grid;
  grid-template-columns: repeat(3, auto);
  grid-column: 1 / 2;
}

main#game > header ul li {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 4px;
}

main#game > header ul li:first-of-type {
  justify-self: end;
}

main#game > header ul li:last-of-type {
  justify-self: start;
}

main#game img {
  max-width: 1830px;
  flex: auto;
  cursor: crosshair;
}

section#map {
  position: relative;
}

/* div#target {
  min-width: 50px;
  min-height: 50px;
  border: 2px dashed white;
  border-radius: 50%;
  background-color: rgba(255, 255, 255, 0.2);
  color: red;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: crosshair;
} */
