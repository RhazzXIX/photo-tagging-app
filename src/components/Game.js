import "../styles/Game.css";
import useGameData from "../assists/useGameData";
import { useParams } from "react-router-dom";
import { useEffect, useState } from "react";

export default function Game(props) {
  const { gameVer } = useParams();
  const [showLoading, setShowLoading] = useState(true);
  const [loading, setLoading] = useState("Loading.");

  const { firstMap, scndMap, charSelection1, charSelection2 } = useGameData();

  let map, charSelection;

  switch (gameVer) {
    case "animeX":
      map = firstMap;
      charSelection = charSelection1;
      break;
    case "gameX":
      map = scndMap;
      charSelection = charSelection2;
      break;
    default:
      map = null;
      charSelection = null;
  }

  const editLoading = () => {
    switch (loading) {
      case "Loading.":
        setLoading("Loading..");
        break;
      case "Loading..":
        setLoading("Loading...");
        break;
      default:
        setLoading("Loading.");
    }
  };

  const removeLoading = (e) => {
    setShowLoading(false);
  };

  useEffect(() => {
    const editLoadingInterval = setInterval(() => {
      editLoading();
    }, 1000);
    if (!showLoading) clearInterval(editLoadingInterval);
    return () => {
      clearInterval(editLoadingInterval);
    };
  }, [loading]);

  return (
    <main id="game">
      <header>
        <ul>
          {charSelection &&
            charSelection.map((char, i) => {
              if (i === charSelection.length - 1)
                return (
                  <li key={char.name}>
                    <img
                      src={char.url}
                      alt={char.name}
                      className="chars"
                      onLoad={removeLoading}
                    />
                    <p>{char.name}</p>
                  </li>
                );
              return (
                <li key={char.name}>
                  <img src={char.url} alt={char.name} className="chars" />
                  <p>{char.name}</p>
                </li>
              );
            })}
        </ul>
      </header>
      {map && <img src={map.url} alt={map.name} />}
      {showLoading && (
        <section id="loading">
          <h1>{loading}</h1>
        </section>
      )}
    </main>
  );
}
