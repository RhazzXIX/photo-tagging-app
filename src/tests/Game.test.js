import "@testing-library/jest-dom";
import { screen, render, waitFor, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Game from "../components/Game";
import useGameData from "../assists/useGameData";
import { RouterProvider, createMemoryRouter } from "react-router-dom";

const routes = [
  {
    path: "/game/:gameVer",
    element: <Game />,
  },
];

const fetchedData = {
  firstMap: {
    miniUrl: "mini url 1",
    name: "Anime Crossover",
    url: "url 2",
  },
  scndMap: {
    miniUrl: "mini url 2",
    name: "Game Crossover",
    url: "url 2",
  },
  charSelection1: [
    {
      name: 'Yato "Yaboku"',
      url: "yato url",
      id: 1,
    },
    {
      name: 'ELUCIA"',
      url: "elucia url",
      id: 2,
    },
    {
      name: 'EVANS"',
      url: "evans url",
      id: 3,
    },
  ],
  charSelection2: [
    {
      name: "Spyro",
      url: "spyro url",
      id: 4,
    },
    {
      name: 'Arthur"',
      url: "arthur url",
      id: 5,
    },
    {
      name: 'Marco"',
      url: "marco url",
      id: 6,
    },
  ],
};

const user = userEvent.setup();

jest.mock("../assists/useGameData", () => {
  return () => fetchedData;
});

describe("Game component", () => {
  it("Renders on screen", () => {
    const { container } = render(<Game />);

    expect(container).toMatchSnapshot();
  });
  it("Loading section is removed after the game is loaded", async () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ["/game", "/game/gameX"],
      initialIndex: 1,
    });
    render(<RouterProvider router={router} />);
    expect(
      screen.getByRole("heading", { name: /loading/i })
    ).toBeInTheDocument();
    const marcoImg = screen.getByAltText(/marco/i);

    fireEvent.load(marcoImg);

    expect(
      screen.queryByRole("heading", { name: /loading/i })
    ).not.toBeInTheDocument();
  });

  describe("Loads the selected game version", () => {
    it("Loads the Anime Crossover verions", () => {
      const router = createMemoryRouter(routes, {
        initialEntries: ["/game/animeX"],
      });
      render(<RouterProvider router={router} />);

      const evansImg = screen.getByAltText(/evans/i);

      fireEvent.load(evansImg);

      expect(screen.queryByAltText(/game crossover/i)).not.toBeInTheDocument();

      expect(screen.getByAltText(/anime crossover/i)).toBeInTheDocument();
    });

    it("Loads the Game Crossover verions", () => {
      const router = createMemoryRouter(routes, {
        initialEntries: ["/game/gameX"],
      });
      render(<RouterProvider router={router} />);

      const marcoImg = screen.getByAltText(/marco/i);

      fireEvent.load(marcoImg);

      expect(screen.queryByAltText(/anime crossover/i)).not.toBeInTheDocument();

      expect(screen.getByAltText(/game crossover/i)).toBeInTheDocument();
    });
  });

  it("Starts the time after the game loads", async () => {
    let delay = 0;
    const router = createMemoryRouter(routes, {
      initialEntries: ["/game/gameX"],
    });

    render(<RouterProvider router={router} />);
    const ms = screen.getByText("0ms");
    const marcoImg = screen.getByAltText(/marco/i);
    expect(
      screen.getByRole("heading", { name: /loading/i })
    ).toBeInTheDocument();

    await waitFor(() => {
      if (delay === 0) {
        delay += 1;
        throw new Error("Did not delay");
      }
      expect(ms.textContent).toMatch("0ms");
    });
    fireEvent.load(marcoImg);
    await waitFor(() => {
      if (delay === 1) {
        delay += 1;
        throw new Error("Did not delay");
      }
    });
    expect(
      screen.queryByRole("heading", { name: /loading/i })
    ).not.toBeInTheDocument();
    expect(ms.textContent).not.toMatch("0ms");
  });

  it("Reveals the character buttons when the map is cliked", async () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ["/game/gameX"],
    });
    render(<RouterProvider router={router} />);

    const marcoImg = screen.getByAltText(/marco/i);

    fireEvent.load(marcoImg);

    const mapSection = screen.getByTestId("map");
    user.click(mapSection);

    await waitFor(() => {
      if (screen.getAllByRole("button").length === 0)
        throw new Error("Buttons did not load");
    });

    expect(screen.getByRole("button", { name: /spyro/i })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /arthur/i })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /marco/i })).toBeInTheDocument();
  });
});
