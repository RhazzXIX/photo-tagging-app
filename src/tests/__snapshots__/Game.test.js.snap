// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component Renders on screen 1`] = `
<div>
  <main
    id="game"
  >
    <header>
      <div
        class="timer"
      >
        <p>
          0
          m
        </p>
        <p>
          0
          s
        </p>
        <p>
          0
          ms
        </p>
      </div>
      <ul />
    </header>
    <section
      id="loading"
    >
      <h1>
        Loading.
      </h1>
    </section>
  </main>
</div>
`;
